pkg/tests/end_to_end_tests/multi_tenancy_test.go

```diff
@@ -0,0 +1,383 @@
+					{Timestamp: 2, Value: 0.2},
+					{Timestamp: 3, Value: 0.3},
+					{Timestamp: 4, Value: 0.4},
+				},
+			},
+		}
+
+		result, err := qr.Query(&prompb.Query{
+			Matchers: []*prompb.LabelMatcher{
+				{
+					Type:  prompb.LabelMatcher_EQ,
+					Name:  model.MetricNameLabelName,
+					Value: "firstMetric",
+				},
+				{
+					Type:  prompb.LabelMatcher_EQ,
+					Name:  tenancy.TenantLabelKey,
+					Value: "tenant-a",
+				},
+			},
+			StartTimestampMs: 2,
+			EndTimestampMs:   4,
+		})
+		require.NoError(t, err)
+
+		// Verifying result.
+		verifyResults(t, expectedResult, result)
+
+		// ----- query across multiple tenants (tenant-a & tenant-c) -----
+		expectedResult = []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "secondMetric"},
+					{Name: "job", Value: "baz"},
+					{Name: "ins", Value: "tag"},
+					{Name: tenancy.TenantLabelKey, Value: "tenant-a"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 2.2},
+					{Timestamp: 3, Value: 2.3},
+					{Timestamp: 4, Value: 2.4},
+				},
+			},
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "secondMetric"},
+					{Name: "job", Value: "baz"},
+					{Name: "ins", Value: "tag"},
+					{Name: tenancy.TenantLabelKey, Value: "tenant-c"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 2.2},
+					{Timestamp: 3, Value: 2.3},
+					{Timestamp: 4, Value: 2.4},
+				},
+			},
+		}
+
+		result, err = qr.Query(&prompb.Query{
+			Matchers: []*prompb.LabelMatcher{
+				{
+					Type:  prompb.LabelMatcher_EQ,
+					Name:  model.MetricNameLabelName,
+					Value: "secondMetric",
+				},
+				{
+					Type:  prompb.LabelMatcher_RE,
+					Name:  tenancy.TenantLabelKey,
+					Value: "tenant-a|tenant-c",
+				},
+			},
+			StartTimestampMs: 2,
+			EndTimestampMs:   4,
+		})
+		require.NoError(t, err)
+
+		// Verifying result.
+		verifyResults(t, expectedResult, result)
+	})
+}
+
+func TestMultiTenancyWithValidTenants(t *testing.T) {
+	ts, tenants := generateSmallMultiTenantTimeseries()
+	withDB(t, *testDatabase, func(db *pgxpool.Pool, t testing.TB) {
+		// With valid tenants.
+		cfg := tenancy.NewSelectiveTenancyConfig(tenants[:2], false) // valid tenant-a & tenant-b.
+		mt, err := tenancy.NewMultiTenancy(cfg)
+		require.NoError(t, err)
+
+		// Ingestion.
+		client, err := pgclient.NewClientWithPool(&pgclient.Config{HAEnabled: false}, 1, pgxconn.NewPgxConn(db), mt)
+		require.NoError(t, err)
+
+		// Ingest tenant-a.
+		_, err = client.Ingest(ingstr.Request{Tenant: tenants[0], Ts: ts, Req: ingstr.NewWriteRequest()})
+		require.NoError(t, err)
+		// Ingest tenant-b.
+		_, err = client.Ingest(ingstr.Request{Tenant: tenants[1], Ts: ts, Req: ingstr.NewWriteRequest()})
+		require.NoError(t, err)
+		// Ingest tenant-c.
+		_, err = client.Ingest(ingstr.Request{Tenant: tenants[2], Ts: ts, Req: ingstr.NewWriteRequest()}) // Invalid ingestion, so this should not get ingested.
```

Comment:
Important! shouldn't this return an error? Trying to ingest an invalid tenant should result in auth error

Reply:
Ah yes, thank you for catching that. The VerifyAndApplyTenant didn't check at start whether the header tenant is authorized. This is fixed in recent commits.